design/design-doc-resource-claim-definition.md

```diff
@@ -0,0 +1,174 @@
+              type: string
+```
+
+```yaml
+---
+apiVersion: core.crossplane.io/v1alpha1
+kind: ResourceBindingDefinition
+metadata:
+  name: mysqlinstance-cloudsqlinstance
+spec:
+  claim:
+    apiVersion: database.crossplane.io/v1alpha1
+    kind: MySQLInstance
+  resources:
+  - resource:
+      apiVersion: database.gcp.crossplane.io/v1beta1
+      kind: CloudSQLInstance
+    class:
+      apiVersion: database.gcp.crossplane.io/v1beta1
+      kind: CloudSQLInstanceClass
+      strategies:
+        provisioning:
+        - overlay
+        - translate
+        overlay:
+        - fromClassField: ".specTemplate"
+          toResourceField: ".spec"
+        translate:
+        - fromClaimField: ".spec.engineVersion"
+          toResourceField: ".spec.forProvider.databaseVersion"
+          transforms:
+            toupper: true
+            replace:
+              string: "."
+              with: "_"
+---
+apiVersion: core.crossplane.io/v1alpha1
+kind: ResourceBindingDefinition
+metadata:
+  name: mysqlinstance-cloudsqlinstance
+spec:
+  claim:
+    apiVersion: database.crossplane.io/v1alpha1
+    kind: MySQLInstance
+  resources:
+  - resource:
+      apiVersion: database.azure.crossplane.io/v1beta1
+      kind: MySQLServer
+    class:
+      apiVersion: database.azure.crossplane.io/v1beta1
+      kind: SQLServerClass
+      strategies:
+        selection:
+        - match
+        match:
+        - fromClaimField: ".spec.engineVersion"
+          toClassField: ".specTemplate.forProvider.version"
+          transforms:
+            toupper: true
+  - resource:
+      apiVersion: database.azure.crossplane.io/v1alpha3
+      kind: MySQLServerVirtualNetworkRule
+    class:
+      apiVersion: database.azure.crossplane.io/v1alpha3
+      kind: MySQLServerVirtualNetworkRuleClass
+```
+
+`ResourceClaimDefinition` is an abstraction that provides guard rails and
+opinions around the process of authoring a `CustomResourceDefinition` that is
+intended for use as a resource claim. Crossplane resource claims are of many
+different kinds, but all have certain things in common. These include:
+
+* Declaring use of the status subresource.
+* Supporting standard Kubernetes object and type metadata.
+* Supporting a binding phase and conditioned status in their `.status`.
+* Supporting a class selector, class reference, resource reference, and
+  connection secret reference in their `.spec`.
+* Supporting a standard set of kubectl printer columns.
+
+A `ResourceClaimDefinition` allows an infrastructure operator to define a
+resource claim without having to specify all of these things correctly. They can
+instead focus on what they're concerned with - defining the purpose specific
+`.spec` fields of their claim. A `ResourceClaimDefinition` controller will watch
+for `ResourceClaimDefinition` instances, filling in the blanks to create a
+corresponding `CustomResourceDefinition`.
+
+`ResourceBindingDefinition` maps a kind of resource claim to the kinds of
```

Comment:
When an `XClaim` is created, the only selection we normally do is deciding which `XClass` to refer. For generic `ResourceClaim`, we need to have a mechanism to select which `ResourceBindingDefinition` to use. It'd be helpful to mention how you approach that problem.

Reply:
I don't think we'd select one as such. I imagine it would be more using the set of RBDs as constraints when selecting a class. The RBDs constrain which resource classes are eligible for static and dynamic provisioning.